STYLE_GUIDE.md

```diff
@@ -0,0 +1,1245 @@
+
+<details>
+<summary>Example</summary>
+
+#### Good
+
+```hrx
+<===> single/input.scss
+a {
+  b {c: d}
+}
+
+<===> single/output.css
+a b {
+  c: d;
+}
+
+<===>
+================================================================================
+<===> double/input.scss
+a {
+  b, c {d: e}
+}
+
+<===> double/output.css
+a b, a c {
+  d: e;
+}
+```
+
+#### Bad
+
+```hrx
+<===> single/input.scss
+a {
+  b {c: d}
+}
+
+<===> single/output.css
+a b {
+  c: d;
+}
+
+<===> double/input.scss
+a {
+  b, c {d: e}
+}
+
+<===> double/output.css
+a b, a c {
+  d: e;
+}
+```
+
+</details>
+
+The first file of each spec within an HRX file other than the first should begin with:
+
+```hrx
+<===>
+================================================================================
+```
+
+This is an HRX comment, and has no effect on the semantics of the file. It just
+serves to visually separate specs from one another.
+
+These comment separators should also be added before specs' parent directories
+if they're added separately, such as for an `options.yml` or [a `README.md`][]
+file.
+
+[a `README.md`]: #consider-adding-readme-md-files-to-parent-directories
+
+### DO order files consistently
+
+<details>
+<summary>Example</summary>
+
+#### Good
+
+```hrx
+<===> top_level/options.yml
+---
+:warning_todo:
+- libsass # sass/libsass#2834
+:ignore_for:
+- ruby-sass
+
+<===> top_level/input.scss
+$var: value !global;
+a {b: $var}
+
+<===> top_level/output.css
+a {
+  b: value;
+}
+
+<===> top_level/warning
+DEPRECATION WARNING: As of Dart Sass 2.0.0, !global assignments won't be able to
+declare new variables. Consider adding `$var: null` at the top level.
+
+  ,
+1 | $var: value !global;
+  | ^^^^^^^^^^^^^^^^^^^
+  '
+    /sass/spec/variables/global/first_declaration/top_level/input.scss 1:1  root stylesheet
+```
+
+#### Bad
+
+```hrx
+<===> top_level/output.css
+a {
+  b: value;
+}
+
+<===> top_level/warning
+DEPRECATION WARNING: As of Dart Sass 2.0.0, !global assignments won't be able to
+declare new variables. Consider adding `$var: null` at the top level.
+
+  ,
+1 | $var: value !global;
+  | ^^^^^^^^^^^^^^^^^^^
+  '
+    /sass/spec/variables/global/first_declaration/top_level/input.scss 1:1  root stylesheet
+
+<===> top_level/options.yml
+---
+:warning_todo:
+- libsass # sass/libsass#2834
+:ignore_for:
+- ruby-sass
+
+<===> top_level/input.scss
+$var: value !global;
+a {b: $var}
+```
+k
+</details>
+
+To keep specs consistent, their constituent files should always be written in
+the following order:
+
+1. `options.yml`.
+2. `input.scss` or `input.sass`.
+3. Any files imported or used by `input.scss` or `input.sass`.
+4. `output.css` or `error`
+5. `warning`
+
+If a spec has [implementation-specific expectations][], those expectations
+should go directly after the corresponding default expectation.
+
+[implementation-specific expectations]: README.md#implementation-specific-expectations
+
+### DO use Dart Sass's output as the default
```

Comment:
Does this also apply to errors? Currently, dart-sass errors are sometimes inferior to ruby and libsass, e.g.:

libsass and ruby:

> Invalid CSS after "... (foo: bar) or ": expected @supports condition (e.g. (display: flexbox)), was "{"

dart-sass:

> expected "("

Reply:
Yes. This isn't about saying which errors are best, it's just about being consistent in formatting. If any implementation's errors need improvement, that's something that should be handled on that implementation's issue tracker, not by swapping output positions in the specs.

Reply:
If `error` is not what we want the canonical error message to look like, but is merely the current error message of dart-sass, perhaps all errors should be language-specific?

I.e. no `error`, only `error-dart-sass` and `error-libsass`.

Reply:
I think that could be a reasonable approach. Let's discuss it on a separate issue, though; until the infrastructure supports it, I'd like to keep this as the recommended approach.